cfgrammar/src/lib/span.rs

```diff
@@ -237,35 +234,61 @@ mod tests {
         }
     }
 
-    fn newline_test_helper(feed: &[&str], tests: &[Vec<(usize, usize)>]) {
-        assert_eq!(feed.len(), tests.len());
+    fn newline_test_helper(feed: &[&str], tests: &[(usize, usize)]) {
         let mut cache = NewlineToLineColCache::default();
         for f in feed {
             cache.feed(f);
         }
 
         let mut src_pos = 0;
-        for (f, test) in feed.iter().zip(tests) {
-            let mut result = Vec::new();
+        let mut result = Vec::new();
+        for f in feed {
             for (offset, _) in f.char_indices() {
                 let line_col = cache.byte_to_line_and_col(src_pos, f, offset + src_pos);
                 result.push(line_col.unwrap())
             }
             src_pos += f.len();
-            assert_eq!(&result, test)
         }
+        assert_eq!(result, tests)
     }
 
     #[test]
-    fn all_one_line() {
-        newline_test_helper(
-            &["a", "b", "c"],
-            &[vec![(1, 1)], vec![(1, 2)], vec![(1, 3)]],
-        );
-        newline_test_helper(&["abc"], &[vec![(1, 1), (1, 2), (1, 3)]]);
-        newline_test_helper(&["ab", "c"], &[vec![(1, 1), (1, 2)], vec![(1, 3)]]);
-        newline_test_helper(&["a\n", "b"], &[vec![(1, 1), (1, 2)], vec![(2, 1)]]);
-        newline_test_helper(&["a", "\nb"], &[vec![(1, 1)], vec![(1, 2), (2, 1)]]);
+    fn newline_cache_various() {
+        let abc = &[(1, 1), (1, 2), (1, 3)];
+        newline_test_helper(&["a", "b", "c"], abc);
+        newline_test_helper(&["abc"], abc);
+        newline_test_helper(&["ab", "c"], abc);
+
+        let ab = &[(1, 1), (1, 2), (2, 1)];
+        newline_test_helper(&["a\n", "b"], ab);
+        newline_test_helper(&["a", "\nb"], ab);
+
+        let cr = &[(1, 1), (1, 2), (1, 3), (2, 1)];
+        newline_test_helper(&["a\r\n", "b"], cr);
+        newline_test_helper(&["a", "\r\nb"], cr);
+        newline_test_helper(&["a\r", "\nb"], cr);
```

Comment:
So, the lack of incrementing trailing chars removed in this patch is because,
I kind of feel like this should test should be for `&[(1, 1), (1, 2), (1, 2), (2, 1)]` rather than `... (1, 2), (1, 3), ..`

But doing it uniformly across split strings is a bit of work, but we should decide what we want the output to be here.
With special casing `\r` in feed, it was giving `&[.., (1, 2), (1, 2), ...]` for the 3rd `cr` case, but `(1, 2), (1, 3)` for the first 2.

Any thoughts?

Reply:
Personally I'm fine with just ignoring `\r` when calculating newlines and treating it as a zero width character for column positions. It's simple and has the same effect (AFAICT) as "complicated" solutions.

Reply:
Alright, that is going to take some work though.

Reply:
Well, I'd tried ignoring `\r`, but ended up ignoring '\n' if '\n' is preceded by '\r', primarily because,
if '\r' appears at the beginning of a line, we could end up with a `column: 0`, which didn't seem right.

Reply:
I believe this is fixed in [21d2f1f](https://github.com/softdevteam/grmtools/pull/299/commits/21d2f1ffac040002ba90eecd49b09b2140145f5a).